### PR TITLE
Working prototype

### PR DESCRIPTION
**Pull request by brtkwr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+OpenStack Container Clusters
+============================
+
+This role can be used to register container cluster templates in Magnum
+using the Magnum CLI.
```
Comment (markgoddard) [inline README.md]: Needs updating

File: README.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+Requirements
+------------
+
+The OpenStack Magnum API should be accessible from the target host.
+
+Role Variables
+--------------
+
+`os_container_infra_config.endpoint` is the endpoint for Magnum API.
```
Comment (markgoddard) [inline README.md]: Typically in ansible we use flat variables (e.g. ``os_container_infra_config_endpoint``), because they can be given defaults, and overridden individually. With dict type variables, you have to set all the items.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          ---
+          clouds:
+            mycloud:
+              auth:
+                auth_url: http://openstack.example.com:5000
+                project_name: p3
+                username: user
+                password: secretpassword
+              region: RegionOne
+          my_container_infra_config: |
```
Comment (markgoddard) [inline README.md]: This is odd. Why define this as YAML in YAML? Shouldn't my_container_infra_config just be a variable?

File: tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    cluster_template_name: "{{ os_container_infra_config.cluster_template_name }}"
+    master_count: "{{ os_container_infra_config.master_count }}"
+    node_count: "{{ os_container_infra_config.node_count }}"
+    keypair: "{{ os_container_infra_config.keypair }}"
+  register: os_container_infra_result
+  become: no
+
+- debug: var=os_container_infra_result
+
+- fail:
```
Comment (markgoddard) [inline tasks/main.yml]: erm

File: tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    magnum_endpoint: "{{ os_container_infra_config.magnum_endpoint }}"
+    cluster_name: "{{ os_container_infra_config.cluster_name }}"
+    cluster_template_name: "{{ os_container_infra_config.cluster_template_name }}"
+    master_count: "{{ os_container_infra_config.master_count }}"
+    node_count: "{{ os_container_infra_config.node_count }}"
+    keypair: "{{ os_container_infra_config.keypair }}"
+  register: os_container_infra_result
+  become: no
+
+- debug: var=os_container_infra_result
```
Comment (markgoddard) [inline tasks/main.yml]: Probably don't need to leave this in

File: tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+---
+
+- name: Get magnum access info
```
Comment (markgoddard) [inline tasks/main.yml]: Isn't this creating a cluster?

File: tests/test.yml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+---
+- hosts: localhost
+  connection: local
+  remote_user: root
```
Comment (markgoddard) [inline tests/test.yml]: Why root?

File: library/os_container_infra.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            cluster_template_name=dict(required=True, type='str'),
+            master_count=dict(required=True, type='int'),
+            node_count=dict(required=True, type='int'),
+            keypair=dict(required=True, type='str'),
+        ),
+        supports_check_mode=False
+    )
+
+    cloud_name = module.params['cloud_name']
+    magnum_endpoint = module.params['magnum_endpoint']
```
Comment (markgoddard) [inline library/os_container_infra.py]: This shouldn't be required - we should be able to use the keystone endpoint catalogue to look up the magnum endpoint. Does it not work without endpoint_override?

File: library/os_container_infra.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    os_cloud = get_os_cloud(cloud_name,  module)
+    magnum_client = Client('1', endpoint_override=magnum_endpoint, session=os_cloud.session)
+
+    cluster_template_name = module.params['cluster_template_name']
+    cluster_template_id = magnum_client.cluster_templates.get(cluster_template_name).uuid
```
Comment (markgoddard) [inline library/os_container_infra.py]: Failures need to be handled by module.fail_json() - what if the template doesn't exist?

File: library/os_container_infra.py [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+                    node_count=module.params['node_count'],
+                    keypair=module.params['keypair'],
+                )
+
+    tries = 0
+    fail_after = 100
+    changed = False
+    while True:
+        container_infra_details = get_update_or_create_container_infra(magnum_client, module, **kwargs)
+        display.warning(container_infra_details)
```
Comment (markgoddard) [inline library/os_container_infra.py]: Why are we warning here? Can't you use module.debug()?

File: tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+  os_container_infra:
+    cloud_name: "{{ os_container_infra_config.cloud_name }}"
+    magnum_endpoint: "{{ os_container_infra_config.magnum_endpoint }}"
+    cluster_name: "{{ os_container_infra_config.cluster_name }}"
+    cluster_template_name: "{{ os_container_infra_config.cluster_template_name }}"
+    master_count: "{{ os_container_infra_config.master_count }}"
+    node_count: "{{ os_container_infra_config.node_count }}"
+    keypair: "{{ os_container_infra_config.keypair }}"
+  register: os_container_infra_result
+  become: no
```
Comment (markgoddard) [inline tasks/main.yml]: This is the default

File: library/os_container_infra.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+                # Everything looks okay, return cluster details
+                return container_infra_details
+        else:
+            module.fail_json(
+                msg="Cluster returned unexpected status: {}".format(container_infra_details.status))
+    except NotFound:
+        # Create new cluster here if no cluster with the name was found
+        magnum_client.clusters.create(**kwargs)
+        display.warning("Cluster called {} does not exist, creating new.".format(kwargs['name']))
+        return False
```
Comment (markgoddard) [inline library/os_container_infra.py]: None might be more typical here - then the type is a bit better defined.

File: library/os_container_infra.py [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            cloud_name=dict(required=True, type='str'),
+            magnum_endpoint=dict(required=True, type='str'),
+            cluster_name=dict(required=True, type='str'),
+            cluster_template_name=dict(required=True, type='str'),
+            master_count=dict(required=True, type='int'),
+            node_count=dict(required=True, type='int'),
+            keypair=dict(required=True, type='str'),
```
Comment (markgoddard) [inline library/os_container_infra.py:199]: It's common to add a state argument, with a default of 'present' and possible value 'absent'. This allows you to also delete the cluster using ansible, which can be useful.

File: library/os_container_infra.py [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+                )
+
+    tries = 0
+    fail_after = 100
+    changed = False
+    while True:
+        container_infra_details = get_update_or_create_container_infra(magnum_client, module, **kwargs)
+        display.warning(container_infra_details)
+        if container_infra_details:
+            stack_id = container_infra_details.stack_id
```
Comment (markgoddard) [inline library/os_container_infra.py]: Wouldn't the cluster ID be more useful than the stack_id?

File: library/os_container_infra.py [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+            display.warning(stack_id)
+            module.exit_json(changed=changed, details=stack_id)
```
Comment (markgoddard) [inline library/os_container_infra.py]: Why not return all the cluster details?

File: library/os_container_infra.py [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+import time
+
+display = Display()
+
+def get_os_cloud(cloud_name, module):
+    os_cloud = openstack.connect(cloud=cloud_name)
+    os_cloud.authorize()
+    return os_cloud
+
+def get_update_or_create_container_infra(magnum_client, module, **kwargs):
```
Comment (markgoddard) [inline library/os_container_infra.py]: **No need to change**, but the swiss army knife function approach can be a little hard to follow, and means that every poll iteration has the possibility of creating or updating the cluster, when it should just be querying. Consider something like this, with the logic factored out to the top level:

* Does the cluster exist?
* If we want it to exist:
  * If it does not exist, create and wait for completion or failure
  * If it does exist, update and wait for completion or failure
* If we do not want it to exist:
  * If it does exist, delete it and wait for deletion

File: README.md [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+`os_container_infra_config.keypair` is the keypair to use to access the cluster nodes.
+
+`os_container_infra_config.master_count` is the number of master nodes.
+
+`os_container_infra_config.node_count` is the number of worker nodes.
+
+`os_container_infra_config.interfaces` is a list of additional network interfaces to
+attach to the cluster since Magnum only allows attachment of one network
+interface by default.
```
Comment (markgoddard) [inline README.md]: I don't think we're using this here.

File: README.md [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+Example Playbook
+----------------
+
+The following playbook registers a cluster template.
```
Comment (markgoddard) [inline README.md]: cluster. It also creates clouds.yml

**Review by markgoddard — CHANGES_REQUESTED** [review]
Thanks Bharat

**Comment by markgoddard** [comment]
I just enabled TravisCI for this repo.

File: README.md (hunk 3 above)
Comment (brtkwr) [reply README.md]: Its a new paradigm: yaml-in-yaml!

File: library/os_container_infra.py (hunk 10 above)
Comment (brtkwr) [reply library/os_container_infra.py]: module.debug() does not appear to fire any messages that are visible. on the contrary, display.warning can be grepped inside ansible.log.

File: library/os_container_infra.py (hunk 16 above)
Comment (brtkwr) [reply library/os_container_infra.py]: I agree, it is the first iteration and this function is a bit of a mess!

File: library/os_container_infra.py (hunk 14 above)
Comment (brtkwr) [reply library/os_container_infra.py]: we want to go from cluster name->cluster inventory. as a result, i think stack_id is the next best place to look.

File: library/os_container_infra.py (hunk 10 above)
Comment (markgoddard) [reply library/os_container_infra.py]: The problem is though, that these aren't actually warnings. If the user configured logging, they would wonder what the problem was.

File: README.md (hunk 17 above)
Comment (brtkwr) [reply README.md]: not yet

File: library/os_container_infra.py (hunk 10 above)
Comment (markgoddard) [reply library/os_container_infra.py]: Here's what I did in the drac module: https://github.com/stackhpc/drac/blob/master/library/drac.py#L641. I don't know if the approach would still work.

**Comment by brtkwr** [comment]
I still see "Continuous integration has not been set up" at the bottom

**Comment by markgoddard** [comment]
You might need to push more changes to trigger a build.

File: library/os_container_infra.py [hunk 19]
```diff
@@ -114,23 +110,46 @@ def main():
...
                 changed = True
+                container_infra = False
+
+    stack_id = container_infra.stack_id
+    display.debug('[DEBUG] Value of stack_id: {}'.format(stack_id))
+
+    stack_resources = get_stack_resources(stack_id)
+
+    heat_client = HeatClient('1', session=os_cloud.session)
+    resources = heat_client.resources.list(stack_id,nested_depth=4)
```
Comment (markgoddard) [inline library/os_container_infra.py]: Do we need to do anything with heat in this module? Why not keep it simple, and limited to create/update/delete a magnum cluster? Any heat queries we need could be done as a separate task, given the stack ID.

Comment (brtkwr) [reply library/os_container_infra.py]: I was going to bundle up everything associated with the cluster back to the user.

File: tests/test.yml (hunk 7 above)
Comment (brtkwr) [reply tests/test.yml]: Shouldn't be root

File: library/os_container_infra.py (hunk 13 above)
Comment (brtkwr) [reply library/os_container_infra.py:199]: I added a state variable in argument_spec

**Comment by markgoddard** [comment]
Is this ready for review? Doug often adds a 'wip' label if he's still making changes, which is quite helpful.

**Comment by brtkwr** [comment]
I've added the label, currently trying to figure out whether python-openstackclient has a way of looking up network objects using their ID or whether I'm going to have to bring in python-neutronclient as well...

**Comment by brtkwr** [comment]
Okay, this is ready for review @markgoddard

**Comment by markgoddard** [comment]
I think I enabled TravisCI for the wrong repo (container-clusters). I can't seem to enable it for this repo as I don't have admin rights to it. @brtknr could you try enabling it, or else we can ask Stig.

**Comment by brtkwr** [comment]
I don't have the permission to enable CI either. Over to you @oneswig.

File: .travis.yml [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+
+  # Create ansible.cfg with correct roles_path
+  - printf '[defaults]\nroles_path=../' >ansible.cfg
+
+  # Compensate for repo name being different to the role
+  - ln -s stackhpc.os-container-infra/ ../stackhpc.os-container-infra
+
+script:
+  # Basic role syntax check
+  - ansible-playbook tests/test.yml -i tests/inventory --syntax-check
```
Comment (markgoddard) [inline .travis.yml:29]: I think you removed tests/test.yml.

File: README.md [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+- openstacksdk
+- python-magnumclient
+- python-heatclient
+- python-novaclient
+
+Role Variables
+--------------
+
+`os_container_infra_cloud` is the name of the cloud inside cloud.yaml.
+`os_container_infra_user` is the name of the cloud inside cloud.yaml.
```
Comment (markgoddard) [inline README.md]: This is the SSH user

File: library/os_container_infra.py [hunk 22]
```diff
@@ -0,0 +1,208 @@
...
+'''
+
+EXAMPLES = '''
+# Gather facts about all resources under <stack_id>:
+- os_container_infra:
+    cloud: mycloud
+    stack_id: xxxxx-xxxxx-xxxx-xxxx
+    nested_depth: 4
+    filters:
+      resource_type: OS::Nova::Server
```
Comment (markgoddard) [inline library/os_container_infra.py]: I think this is for the other module.

File: library/os_container_infra.py [hunk 23]
```diff
@@ -0,0 +1,208 @@
...
+            keypair=dict(required=True, type='str'),
+        ),
+        supports_check_mode=False
+    )
+
+    display = Display()
+
+    container_infra = ContainerInfra(**module.params)
+
+    changed = container_infra.apply()
```
Comment (markgoddard) [inline library/os_container_infra.py]: Looks like we can raise OpenStackError and ValueError. We need to catch those (or just Exception) and run module.fail_json() to gracefully report the error. You could pass the exception message as the fail_json msg argument.

File: library/os_server_interface.py [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+__metaclass__ = type
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: os_server_interface
+short_description: Create, update and delete container infra.
```
Comment (markgoddard) [inline library/os_server_interface.py]: Needs updating

File: library/os_server_interface.py [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+- os_container_infra:
+    cloud: mycloud
+    state: present
+    server_id: xxxxx-xxxxx-xxxx-xxxx
+    interfaces:
+    - p3-lln
+    - p3-bdn
+  register: result
+- debug:
+    var: result['server_interface']
```
Comment (markgoddard) [inline library/os_server_interface.py]: Wrong module again!

File: library/os_server_interface.py [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+     description:
+       - Must be `present` or `absent`.
+     type: str
+   server_id:
+     description:
+        - Server name or uuid.
+     type: str
+   interfaces:
+     description:
+        - List of network interfaces.
```
Comment (markgoddard) [inline library/os_server_interface.py]: The naming here is a little confusing. From the code it looks like this is a list like this:

```
- id: <net ID1>
- id: <net ID2>
```

I'd either just call this argument `networks`, or use `net_id` instead of `id`.

File: library/os_stack_facts.py [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+EXAMPLES = '''
+# Gather facts about all resources under <stack_id>:
+- os_stack_facts:
+    cloud: mycloud
+    stack_id: xxxxx-xxxxx-xxxx-xxxx
+    nested_depth: 4
+    filters:
+      resource_type: OS::Nova::Server
+- debug:
+    var: stack_facts
```
Comment (markgoddard) [inline library/os_stack_facts.py:61]: This one's right!

File: library/os_stack_facts.py [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+        stack_list = self.client.resources.list(
+            stack_id=self.stack_id,
+            nested_depth=self.nested_depth,
+        )
+        result = list()
+        for item in stack_list:
+            item_dict = item.to_dict()
+            condition = True
+            for k,v in self.filters.items():
+                condition = condition and item_dict[k] == v
```
Comment (markgoddard) [inline library/os_stack_facts.py:94]: FYI (no need to change), you can do things like this in Ansible using Jinja filters, e.g.

`servers: {{ stack_facts | selectattr('resource_type', 'equalto', 'OS::Nova::Server') | list }}`

File: library/os_stack_facts.py [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+            filters=dict(default=dict(), type='dict'),
+        ),
+        supports_check_mode=False
+    )
+
+    display = Display()
+
+    stack_facts = StackFacts(**module.params)
+
+    module.exit_json(changed=False,ansible_facts=dict(stack_facts=stack_facts.get()))
```
Comment (markgoddard) [inline library/os_stack_facts.py]: Needs failure handling.

File: library/os_server_interface.py [hunk 30]
```diff
@@ -0,0 +1,126 @@
...
+            interfaces=dict(default=[], type='list'),
+        ),
+        supports_check_mode=False
+    )
+
+    display = Display()
+
+    server_interface = ServerInterface(**module.params)
+
+    changed = server_interface.apply()
```
Comment (markgoddard) [inline library/os_server_interface.py]: Needs failure handling.

File: tasks/main.yml [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+
+- name: Attach interfaces to servers
+  os_server_interface:
+    cloud: "{{ os_container_infra_cloud }}"
+    server_id: "{{ item['physical_resource_id'] }}"
+    interfaces: "{{ os_container_infra_interfaces }}"
+    state: present
+  register: server_interface_facts
+  with_items:
+  - "{{ stack_facts }}"
```
Comment (markgoddard) [inline tasks/main.yml]: nit: This is already a list

File: tasks/main.yml [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+
+- name: Generate cluster inventory
+  template:
+    src: container_infra_inventory.j2
+    dest: ./container_infra_inventory
```
Comment (markgoddard) [inline tasks/main.yml]: cluster-infra makes this location configurable: https://github.com/stackhpc/ansible-role-cluster-infra/blob/master/tasks/main.yml#L30

and defaults to `{{ playbook_dir }}/inventory_{{ cluster_name }}`.

File: tasks/main.yml [hunk 33]
```diff
@@ -0,0 +1,34 @@
...
+  os_container_infra:
+    cloud: "{{ os_container_infra_cloud }}"
+    state: "{{ os_container_infra_state }}"
+    cluster_name: "{{ os_container_infra_cluster_name }}"
+    cluster_template_name: "{{ os_container_infra_cluster_template_name }}"
+    master_count: "{{ os_container_infra_master_count }}"
+    node_count: "{{ os_container_infra_node_count }}"
+    keypair: "{{ os_container_infra_keypair }}"
+
+- name: Get stack facts
```
Comment (markgoddard) [inline tasks/main.yml]: You could use a `block:` to only apply the following tasks when `os_container_infra_state` is `present` (as seen in cluster-infra).

File: templates/container_infra_inventory.j2 [hunk 34]
```diff
@@ -0,0 +1,16 @@
+[openstack]
+localhost   ansible_connection='local'    ansible_python_interpreter='/usr/bin/env python'
+
+[master]
+{% for fact in server_interface_facts.results %}
+{% if fact.item.resource_name is search("-master$") %}
+{{ fact.item.physical_resource_id }}  ansible_host={{ fact.server_interface[os_container_infra_public_interface][-1] }} ansible_user={{ os_container_infra_user }}
```
Comment (markgoddard) [inline templates/container_infra_inventory.j2]: Let's use resource_name here, it's a bit friendlier than the UUID.

File: README.md [hunk 35]
```diff
@@ -0,0 +1,94 @@
...
+      os_container_infra_state: present
+      os_container_infra_cluster_name: test-cluster
+      os_container_infra_cluster_template_name: swarm-fedora-atomic-27
+      os_container_infra_keypair: default
+      os_container_infra_master_count: 1
+      os_container_infra_node_count: 1
+      os_container_infra_public_interface: p3-internal
+      os_container_infra_interfaces:
+      - p3-lln
+      - p3-bdn
```
Comment (markgoddard) [inline README.md]: This is more like how I'd expect this to look, rather than requiring an `id` key. (See later comment for context).

**Review by markgoddard — COMMENTED** [review]
Despite the comments, this is looking pretty good in terms of the code.

File: .travis.yml (hunk 20 above)
Comment (brtkwr) [reply .travis.yml:29]: Oops, I forgot this is why I had it there!

File: README.md [hunk 36]
```diff
@@ -17,15 +17,16 @@ Role Variables
 --------------
 
 `os_container_infra_cloud` is the name of the cloud inside cloud.yaml.
-`os_container_infra_user` is the name of the cloud inside cloud.yaml.
+`os_container_infra_ssh_user` is the name of the cloud inside cloud.yaml.
```
Comment (markgoddard) [inline README.md]: I was referring to the description here.

File: templates/inventory-container-infra.j2 [hunk 37]
```diff
@@ -4,13 +4,13 @@ localhost   ansible_connection='local'    ansible_python_interpreter='/usr/bin/e
 [master]
 {% for fact in server_interface_facts.results %}
 {% if fact.item.resource_name is search("-master$") %}
-{{ fact.item.physical_resource_id }}  ansible_host={{ fact.server_interface[os_container_infra_public_interface][-1] }} ansible_user={{ os_container_infra_user }}
+{{ fact.server_name }}  ansible_ssh_host={{ fact.server_networks[os_container_infra_external_interface][-1] }} ansible_ssh_user={{ os_container_infra_ssh_user }}
```
Comment (markgoddard) [inline templates/inventory-container-infra.j2]: Right first time - ansible_host/ansible_user is the modern way.

**Review by markgoddard — COMMENTED** [review]
Looks good, just a couple of minor issues.

**Review by markgoddard — APPROVED** [review]
Boom

**Comment by brtkwr** [comment]
Okay, once this is merged, I shall integrate this into p3-appliances.

**Comment by brtkwr** [comment]
And also Ansible galaxy?
